connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java

```diff
@@ -434,50 +516,93 @@ public Converter newInternalConverter(boolean isKey, String className, Map<Strin
+                // Attempt to load with the plugin class loader, which uses the current classloader as a fallback
+                String classOrAlias = config.getClass(classPropertyName).getName();
                 try {
-                    klass = pluginClass(
-                            delegatingLoader,
-                            converterClassOrAlias,
-                            HeaderConverter.class
-                    );
+                    klass = pluginClass(delegatingLoader, classOrAlias, basePluginClass, range);
                 } catch (ClassNotFoundException e) {
                     throw new ConnectException(
-                            "Failed to find any class that implements HeaderConverter and which name matches "
-                                    + converterClassOrAlias
-                                    + ", available header converters are: "
-                                    + pluginNames(scanResult.headerConverters())
+                            "Failed to find any class that implements " + basePluginClassName + " and which name matches "
+                                    + classOrAlias + ", available plugins are: "
+                                    + pluginNames(availablePlugins)
                     );
                 }
+                break;
         }
         if (klass == null) {
-            throw new ConnectException("Unable to initialize the HeaderConverter specified in '" + classPropertyName + "'");
+            throw new ConnectException("Unable to initialize the '" + basePluginClassName
+                    + "' specified in '" + classPropertyName + "'");
         }
 
-        String configPrefix = classPropertyName + ".";
-        Map<String, Object> converterConfig = config.originalsWithPrefix(configPrefix);
-        converterConfig.put(ConverterConfig.TYPE_CONFIG, ConverterType.HEADER.getName());
-        log.debug("Configuring the header converter with configuration keys:{}{}", System.lineSeparator(), converterConfig.keySet());
-
-        HeaderConverter plugin;
+        U plugin;
         try (LoaderSwap loaderSwap = withClassLoader(klass.getClassLoader())) {
             plugin = newPlugin(klass);
-            plugin.configure(converterConfig);
+            DefaultArtifactVersion pluginVersion = new DefaultArtifactVersion(PluginScanner.versionFor(plugin));
+            if (range != null && range.hasRestrictions() && !range.containsVersion(pluginVersion)) {
+                // this can happen if the current class loader is used
+                // if there are version restrictions then this should be captured, and we should load using the plugin class loader
+                if (classLoaderUsage == ClassLoaderUsage.CURRENT_CLASSLOADER) {
+                    return getVersionedPlugin(config, classPropertyName, versionPropertyName, basePluginClass, ClassLoaderUsage.PLUGINS, availablePlugins);
+                }
+            }
```

Comment:
Can we assert `range == null || classLoaderUsage == PLUGINS` earlier, to ensure that we never get into this situation where we loaded a class of an invalid version?

I think having this depend on range.containsVersion(pluginVersion) could lead to some strange behavior in the following situation:
* key.converter.version = [1.0, 2.0)
* Converter v1.0 is on the classpath
* Converter v1.2 is in a different plugin
* Converter v1.1 is in the local plugin

If you don't have 1.1 installed, 1.2 is selected. If you later install 1.1 (an older version), you get rolled back to 1.1.
But if you had specified [1.2] it wouldn't get rolled back.

Reply:
Yes, the situation you mentioned is exactly the reason why the public methods to create a converter with a version provided we don't have the option to pass in a class loader and defaults to using the plugins loader, otherwise it can potentially shadow a higher version of a plugin requirement. This code path should ideally never be executed. Will remove this and add the assertion suggestion.